isomerpages/isomercms-frontend
# ref(formField): adds formInput and refactors formField

Part of #771

## Problem
The `FormField` component currently has too much responsibilities. It's setting the title/description etc and has many confusing props, some of which are unused.

After #753, we have more composable form components, which meant that it was possible to refactor this component more cleanly and make it a simpler and easier to use component.

## Solution
The solution outlined here consists of 3 sections.

**remove unused/unneeded props**
the `placeholder` prop, which is given in the [default input api](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input) is preferred over the `defaultValue` prop (which is also never used). 

`maxWidth`/`className` serve the same purpose (setting some css styling on the component) and have been combined into the `className` prop, which is more general. This included 1 new `classNames` package, which allows for easier classname combination.

some oft repeated props (`autocomplete = "off", type = "text"`) were shifted into a base `FormInput` component.

**reusing basic form level components**
The previously defined `FormXxx` components in the form folder was used here to lend semantic understanding to the component. A slight change was also made to separate out the `FormInput` from the `FormField` (this might be superfluous now after the refactor).

The `FormField` component should **not** take a `register` prop as consumers often use it like so: `{...register(id)}`, so a deliberate design choice was made to accept the return values of `register` instead.

**conforming to input spec**
Because the `FormField` is a visual component that wraps an input, we should conform as closely as possible to how an `input` behaves and have minimal additional props on top (or at least hide them). This is in line with the reasoning above in not passing a `register` prop.

**Breaking Changes**



  - Details ...

## Before & After Screenshots

there was a visual change made in `FormMediaInput` and also, an odd lack of the input being `maxWidth`, which will be illustrated below.

### FormMediaInput
**BEFORE**:
![image](https://user-images.githubusercontent.com/44049504/154199153-25b8555a-3735-4e6d-8ca1-dde1f664485e.png)

**AFTER**:
![image](https://user-images.githubusercontent.com/44049504/154199118-fc726d5b-2fb3-4e6a-8e58-8f47ae89e7b6.png)

(technically the slight double line is during the error state because of this statement: ```className={hasError ? `${elementStyles.error}` : "border-0"``` but due to the difficulty of throwing an error in prod, i chose to just use the default state instead.

### FormField's `maxWidth`
the `maxWidth` prop should make the input take all the width available but it seems to have no effect (and has been removed).
 
**dev**
<img width="272" alt="image" src="https://user-images.githubusercontent.com/44049504/154199501-97647c43-cc3d-447f-a83a-cf712aeeb2f1.png">

note that this is with the prop omitted.

**prod**
<img width="207" alt="image" src="https://user-images.githubusercontent.com/44049504/154199558-1a59b0fc-ce37-4931-b9ad-9da2c76c4278.png">

This is with the `maxWidth` prop

## Tests
Due to the huge amount of refactored components, it was not feasible to manually check each and every one of them. The ones that have been checked are those with interesting behaviour. There was no visual breakage on those checked (in the success state). 

For those that were not manually checked, a simple code inspection was done to ensure that the logical component is still the same.

Testing steps: 
1. Open up the CMS 
2. navigate to your localhost
3. open a page containing the component
4. verify that the title is still in the same place. 
5. verify that the error message still shows
6. verify that with no input, the placeholder is still identical
6. verify that the behaviour is still identical 

- [x] directorysettingsmodal
- [x] Hyperlinkmodal
- [x] mediasettings
- [x] pagesettings
- [x] Resourceformfields
- [x] contactcard
- [x] Contactfields
- [x] Locationcard
- [x] locationfields 
- [x] hero button 
- [x] hero dropdown
- [x] herosection 
- [x] Infobar section 
- [x] Info pic section
- [x] keyhighlight 
- [x] resources section	
- [x] media alt text
- [x] nav section
- [x] nav sublink section	
- [x] Editcontactus
- [x] Resourceroom

**Notes**
1. having `onFieldChange` in `FormContext` seems superfluous (as usually each controllable component will have their own controller/handler) and it seems like a mistake to place it there. As we are migrating over to the design system soon, i think it's still acceptable for now to leave it (a band-aid) as design system will not have a `onFieldChange` in context.
2. the difference between `alwaysDisabled` and `isDisabled` lies in how the input component should be used. For wrappers around `FormInput` that **do not** require the input be interactive but only as a visual display of information, `alwaysDisabled` should be true. This is to allow us to have an interactive component, while having the input serve as a source of information, as per `FormFieldMedia`. If we did not have this prop, the input would be interactive, when `isDisabled` is false, which is not desired behaviour
